test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandlerTests.ValidateTokenAsyncTests.Audience.cs

```diff
@@ -0,0 +1,178 @@
+                        TokenAudience = Default.Audience,
+                        TVPAudiences = [Default.Audience + "/"],
+                        ValidationParameters = CreateValidationParameters([Default.Audience + "/"], false),
+                        ExpectedIsValid = false,
+                        ExpectedException = ExpectedException.SecurityTokenInvalidAudienceException("IDX10214:"),
+                        ExpectedExceptionValidationParameters = ExpectedException.SecurityTokenInvalidAudienceException("IDX10215:"),
+                    }
+                };
+
+                static ValidationParameters CreateValidationParameters(
+                    List<string>? audiences,
+                    bool ignoreTrailingSlashWhenValidatingAudience = false)
+                {
+                    ValidationParameters validationParameters = new ValidationParameters();
+                    audiences?.ForEach(audience => validationParameters.ValidAudiences.Add(audience));
+                    validationParameters.IgnoreTrailingSlashWhenValidatingAudience = ignoreTrailingSlashWhenValidatingAudience;
+                    validationParameters.LifetimeValidator = SkipValidationDelegates.SkipLifetimeValidation;
+                    validationParameters.TokenReplayValidator = SkipValidationDelegates.SkipTokenReplayValidation;
+
+                    return validationParameters;
+                }
+            }
+        }
+
+        public class ValidateTokenAsyncAudienceTheoryData : ValidateTokenAsyncBaseTheoryData
+        {
+            public ValidateTokenAsyncAudienceTheoryData(string testId) : base(testId) { }
+
+            public string? TokenAudience { get; internal set; } = Default.Audience;
+
+            public List<string>? TVPAudiences { get; internal set; }
+
+            public Saml2Conditions? Saml2Condition { get; internal set; }
+
+            public bool ignoreTrailingSlashWhenValidatingAudience { get; internal set; } = false;
+        }
+
+        private static Saml2SecurityToken CreateToken(string audience, Saml2Conditions saml2Conditions)
+        {
+            Saml2SecurityTokenHandler saml2TokenHandler = new Saml2SecurityTokenHandler();
+
+            SecurityTokenDescriptor securityTokenDescriptor = new SecurityTokenDescriptor
+            {
+                Expires = DateTime.UtcNow + TimeSpan.FromDays(1),
+                Audience = audience,
+                SigningCredentials = Default.AsymmetricSigningCredentials,
+                Issuer = Default.Issuer,
+                Subject = Default.SamlClaimsIdentity
+            };
+
+            Saml2SecurityToken saml2Token = (Saml2SecurityToken)saml2TokenHandler.CreateToken(securityTokenDescriptor);
+
+            /*
+                    if (saml2Conditions != null)
+                        saml2Token.Assertion.Conditions = saml2Conditions;*/ //TODO: Will adapt this to work with scenarios such as TokenReplay in Jwt or OneTimeUse in SAML.
```

Comment:
Please add a work item or GitHub issue to track the TODO and include the link there. thanks.

Reply:
Will be removing this post my latest sync with Brent. We will we simplifying these tests into only comparing new and old behavior for each validator and have separate tests for these conditions later.